### ISSUE TITLE
Chapel Test Framework

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#9730]
# Test Framework for Chapel

This is an issue intended to explore ideas for Chapel's test framework.

This issue description will be edited as the discussion progresses.

## What are some desired qualities in a test framework?

- Low overhead (minimal boiler plate code) to write tests
    - How much code change is necessary to add a simple test to a library?
- Clear error messages
    - Can I write `assert 1 + 1 == 3` and get an error that tells me the result?
        - e.g. `line 22: AssertError: 2 == 3`
- Easy to identify error messages from failures in order
    - How quickly can the user find the first error message emitted?
- Resiliency to fatal errors
    - Will all tests run even if one test goes off the rails?
- Quick turn-around time in the edit-compile-run cycle
    - Are there significant additional steps beyond compiling and running?
- Robust functionality:
    - Package manager integration
    - Documentation integration
    - Performance testing
    - Test dependency chains
    - Configuration-specific testing
    - Platform-specific testing
    - Specifying multilocality
    - Unit testing in parallel

## Motivation for a test framework

There is no official user-facing Chapel framework, so current options are:

- Writing tests from scratch
    - Not an acceptable solution in this day and age
- `start_test`
    - Intended for developers of the compiler
    - Designed for end-to-end integration and regression testing
    - Not fully documented, and documentation is hard to find
    - Lots of overhead, specifically in number of files
        - requires a separate file for `.good`, at a minimum
- user-developed solutions: [Charcoal](https://github.com/buddha314/Charcoal)
    - Fairly simple in functionality
    - Depends on a lot of language features still in development
    - Takes a Pythonic `unittest` library approach
    - Could provide basis for Chapel's official test framework

## Other languages

Let's look at some other test frameworks:

### [Python: unittest](https://docs.python.org/3/library/unittest.html)

- Object oriented approach
- Inspired by Java's JUnit and similar to many other major unit test frameworks
- Supports test automation, shared setup and shutdown, aggregation of tests
    into collections, and independence of test from reporting framework
- Provides a wide range of testing utility functions, including assertions

An example of what this style might look like in Chapel:

```chpl
use UnitTest;

proc main() {
  // Creates instance of TestCase subclasses and runs their tests
  RunTests();
}

class TestStringMethods: TestCase {

  proc testUpper() {
    this.assertEqual('foo'.upper(), 'FOO')
  }

  proc testIsUpper() {
    this.assertTrue('FOO'.isupper())
    this.assertFalse('Foo'.isupper())
  }

  proc testSplit() {
    s = 'hello world'
    var ss = s.split();
    this.assertEqual(ss, ['hello', 'world'])
  }
}
```

Default output:

```
...
----------------------------------------------------------------------
Ran 3 tests in 0.001s

OK
```

Verbose output:

```
testIsUpper (TestStringMethods) ... ok
testSplit (TestStringMethods) ... ok
testUpper (TestStringMethods) ... ok

----------------------------------------------------------------------
Ran 3 tests in 0.001s

OK
```

Example of failure:

```
..
======================================================================
FAIL: testIsUpper (TestStringMethods)
----------------------------------------------------------------------
File "tests.chpl.", line 32, in testIsUpper
    this.assertEqual('foobar'.upper(), 'FOO')
AssertionError: FOOBAR != FOO
```

**Pros**

- Library would be fully written in Chapel
    - Easily extendable / hackable
- Robust features

**Cons**

- Require significant improvements to the `Reflection`.
- Improvements to stack-tracing
- Enforces class-based testing
    - Can be useful, but also adds some verbosity to simple tests

### [D: UnitTest](https://dlang.org/spec/unittest.html)

- Built-in language feature for tests
- Unit tests are specified by keyword `unittest`
- A `unittest` state is determined by the result of assertions within
- `unittests` are called before the `main()` function of a program
    - Thus, both unittests and the actual program can be run together
- D programs can be compiled with or without unit tests enabled
- Q: How does D handle library dependencies?
  - separate compilation allows turning off `--unittest` in submodules

An example of what this style might look like in Chapel:

```chpl
class Sum
{
  proc add(x, y) { return x + y; }

  unittest {
    var sum = new Sum();
    assert(sum.add(3,4) == 7);
    assert(sum.add(-2,0) == -2);
  }
}
```
- Tests can be modified with global function attributes:
    - This would probably be a `pragma` or `unittest` argument in Chapel

```D
// This is D code
@safe nothrow unittest
{
    auto arr = [1,2,3];
    myFunc(arr);
    assert(arr == [2,2,3]);
}
```

- Tests can be flagged to be included in documentation
    - This could also integrate nicely with `chpldoc`

```chpl
/* Add function */
proc add(x, y) { return x + y; }

/* This code generates an example in the documentation */
unittest
{
  assert(add(1, 1) == 2);
}

/* This code is not generated as an example in the documentation */
private unittest
{
  assert(add(2, 2) == 4);
}
```

**Pros**

- Minimal overhead / boiler plate code
- Fairly simple to learn and use
- Integrates with documentation tool

**Cons**

- Maybe too simple for Chapel's needs
    - Though, this approach could be extended
- Relies on annotations
    - Chapel `pragma` is not intended to be user-facing
    - However, the `unittest` keyword could support arguments or something

### [Rust: test annotations](https://doc.rust-lang.org/book/second-edition/ch11-03-test-organization.html)

- Built-in language feature for tests
- Any function or module can be annotated with `#[cfg(test)]` attribute
- Tests fail upon a `panic` (similar to Chapel `halt`, but less destructive)
- Integrated with Cargo, such that Cargo *is* the test-runner

An example of what this style might look like in Chapel:

```chpl
pragma "test"
module tests {
  pragma "test"
  proc testAdd() {
    assertEqual(1+2, 3);
  }

  pragma "test"
  proc testBadAdd() {
    assertEqual(1+1, 3);
  }
}
```

```bash
$ mason test

running 2 tests
test tests.testBadAdd ... FAILED
test tests.testAdd ... ok

failures:

---- tests.testBadAdd stdout ----
        'tests.TestBadAdd' halted at 'assertion failed: `(left == right)`
  left: `-1`,
  right: `3`', src/lib.chpl:21:8

failures:
    tests.testBadAdd

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

Rust also treats code examples in comment blocks as tests:

````chpl
/*
 First line is a short summary describing function.

 Test code blocks start with triple backquotes:

 ```
 var result = add(2, 3);
 assertEqual(result, 5);
 ```
*/
proc add(a, b) {
  a + b;
}
````

```bash
$ mason test
running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests

running 1 test
test src/lib.chpl - add (line 7) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Pros**

- Minimal overhead / boiler plate code
- Fairly simple to learn and use
- Integrates with documentation tool
- Integrates with package manager / build tool

**Cons**

- Relies heavily on annotations
    - Chapel `pragma` is not intended to be user-facing

## Other Thoughts

- Chapel will likely need a way to provide annotations to tests, whether it be
  through pragmas, a new keyword that takes arguments, or a decorator-like
  feature.

- Configuration-specific testing
    - Tests could introspect Chapel configuration variables and skip depending
        on those
        - This could also be utilized for platform-specific tests

- Specifying multilocality
    - Tests could introspect `numLocales` to determine available locales and run
        skip certain tests depending on that value

- Unit testing in parallel
    - Seems like a reasonable feature to be able to opt into for independent
      single-task tests

- Dependency chains
    - `skipIf` or `dependsOn` conditionals

- Performance testing
    - Performance testing might be separable enough to support as external 
        libraries
    - A basic test framework could support performance testing with logging of
        timings

- How to handle remaining `halts` in the language / standard library
  - This could be solved within the test runner, e.g. `mason test` will resume running after a `halt` is encountered
  - Could run each test inside a subprocess (at the cost of performance)
  - Could compile with these checks turned off, e.g. out of bounds error avoided with `--no-bounds-checking`

**Comment by mppf** [comment]
Swift seems to use a test class approach (like Python unittest): https://developer.apple.com/documentation/xctest

**Comment by mppf** [comment]
About the differing approaches (make test classes like Python unittest or something else) I find that my main reaction is that I really care about tests in documentation comments being recognized. Of the above systems, only Rust does this, but one can certainly imagine converting such tests in documentation into a test class etc etc.

**Comment by lydia-duncan** [comment]
Here's the alternate Python module I mentioned: https://docs.pytest.org/en/latest/

**Comment by tjstavenger-pnnl** [comment]
The current version of JUnit uses annotations in a similar way to Rust. I like the idea of loose coupling between the test code I write and a test framework (i.e., no extending classes or required function overriding, etc.). I think the two biggest things for us are ways to do setUp() and tearDown() to prep the data / environment as well as a suite of assert() functions.

See https://junit.org/junit5/docs/current/user-guide/#writing-tests

**Comment by bradcray** [comment]
Just noting before it's forgotten or lost that over on issue #7495 @russel filed some interesting commentary on Python, D, Groovy, Haskell, etc. that anyone interested in this topic may want to read and/or follow up on (preferable here):

https://github.com/chapel-lang/chapel/issues/7495#issuecomment-407354671
https://github.com/chapel-lang/chapel/issues/7495#issuecomment-407361961

**Comment by ben-albrecht** [comment]
I have submitted a more specific proposal here: https://github.com/chapel-lang/chapel/issues/13098

**Comment by Spartee** [comment]
@ben-albrecht should this issue be closed given that there is a more specific issue and merged unittest PR?

**Comment by ben-albrecht** [comment]
@Spartee -  I consider this issue more of "We need a Chapel unit test framework and here's what we'd like in it", and #13098 to be "Here's a specific design proposal for Chapel's unit test framework". 

I think we should keep both of these open until we have a user-facing unit test framework in master. The recent merge (#13125) is a building block towards this, but we've got a lot of work to do still.

**Comment by ben-albrecht** [comment]
The UnitTest package has been merged into master (https://github.com/chapel-lang/chapel/pull/13806). Further improvements on this package can be opened as separate feature request / design issues.
